File: clojure/pig-latin/src/pig_latin.clj
(ns pig-latin
  (:require [clojure.string :as str]))

(def add-ay-pattern
  "Starts with vowel, or starts with xr/yt"
  #"^([aeiou]|xr|yt)")

(def qu-start-pattern
  "Starts with qu, including with consonants before"
  #"^([^aeiou]*qu)(.*)")

(def y-start-pattern
  "Starts with y after consonants"
  #"^([^aeiou]+)(y.*)")

(def normal-consonant-pattern
  "Starts with consonant, no y/qu special case"
  #"^([^aeiou]+)(.*)")

(defn re-replace-if-match
  [input pattern]
  (when (re-find pattern input)
    (str/replace input pattern "$2$1ay")))

(defn translate-one-word
  [word]
  (cond
    (re-find add-ay-pattern word) (str word "ay")
    :else (->> [qu-start-pattern
                y-start-pattern
                normal-consonant-pattern]
               (keep #(re-replace-if-match word %))
               first)))

(defn translate
  [input]
  (-> input
      (str/split #"\s")
      (->> (map translate-one-word)
           (str/join " "))))

File: python/pig-latin/pig_latin.py
import re

add_ay_pattern = re.compile("^([aeiou]|xr|yt)")
qu_start_pattern = re.compile("^([^aeiou]*qu)(.*)")
y_start_pattern = re.compile("^([^aeiou]+)(y.*)")
consonant_start_pattern = re.compile("^([^aeiou]+)(.*)")

replace_pattern = r'\2\1ay'

def translate_one_word(word):
    if add_ay_pattern.search(word):
        return word + "ay"
    if qu_start_pattern.search(word):
        return re.sub(qu_start_pattern, replace_pattern, word)
    if y_start_pattern.search(word):
        return re.sub(y_start_pattern, replace_pattern, word)
    if consonant_start_pattern.search(word):
        return re.sub(consonant_start_pattern, replace_pattern, word)

def translate(text):
    return " ".join([translate_one_word(word) for word in text.split()])
